File: clients/elixir/lib/eal_client.ex
defmodule EalClient do
  @moduledoc "Runtime-light Embedded Alerts client using OTP's built-in :httpc."

  defstruct [:base_url, :token, headers: []]

  def new(base_url, opts \\ []) do
    %__MODULE__{
      base_url: String.trim_trailing(base_url, "/"),
      token: Keyword.get(opts, :token),
      headers: Keyword.get(opts, :headers, [])
    }
  end

  def request(%__MODULE__{} = client, method, path, body \\ nil) do
    :inets.start()
    :ssl.start()

    headers =
      [{"accept", "application/json"} | client.headers]
      |> maybe_authorize(client.token)
      |> Enum.map(fn {key, value} -> {to_charlist(key), to_charlist(value)} end)

    url = to_charlist(client.base_url <> "/" <> String.trim_leading(path, "/"))

    request =
      if is_nil(body) do
        {url, headers}
      else
        {url, headers, ~c"application/json", IO.iodata_to_binary(body)}
      end

    :httpc.request(method, request, [], body_format: :binary)
  end

  defp maybe_authorize(headers, nil), do: headers
  defp maybe_authorize(headers, token), do: [{"authorization", "Bearer " <> token} | headers]
end

File: clients/ruby/lib/eal_client.rb
# frozen_string_literal: true

require "json"
require "net/http"
require "uri"

class EmbeddedAlertsClient
  def initialize(base_url:, token: nil, headers: {})
    @base_url = base_url.sub(%r{/+$}, "")
    @token = token
    @headers = headers
  end

  def request(method, path, body: nil)
    uri = URI("#{@base_url}/#{path.sub(%r{^/+}, "")}")
    request_class = Net::HTTP.const_get(method.to_s.capitalize)
    req = request_class.new(uri)
    @headers.each { |key, value| req[key] = value }
    req["Authorization"] ||= "Bearer #{@token}" if @token
    if body
      req["Content-Type"] ||= "application/json"
      req.body = JSON.generate(body)
    end

    response = Net::HTTP.start(
      uri.hostname,
      uri.port,
      use_ssl: uri.scheme == "https"
    ) { |http| http.request(req) }

    raise "Embedded Alerts request failed: #{response.code}" unless response.is_a?(Net::HTTPSuccess)
    return nil if response.body.nil? || response.body.empty?

    JSON.parse(response.body)
  end
end
